File: algorithms/aco.go
package algorithms

import (
	"math"
	"math/rand/v2"
	"ant_colony_golang_backend/models"
)

type AntColony struct {
	Points              []models.TurbineFaultPoint
	NAnts               int
	NIterations         int
	NPoints             int
	Pheromone           [][]float64
	Alpha               float64
	Beta                float64
	EvaporationRate     float64
	Q                   float64
	BestPath            []int
	TurbineOrder        []string
	BestPathLength      float64
	BestDowntimeDays    float64
	BestPathLenDowntime float64
	Rnd                 *rand.Rand
}

func NewAntColony(points []models.TurbineFaultPoint, nAnts int, nIterations int, alpha, beta, evaporationRate, q float64, rng *rand.Rand) *AntColony {
	nPoints := len(points)
	pheromone := make([][]float64, nPoints)
	for i := 0; i < nPoints; i++ {
		pheromone[i] = make([]float64, nPoints)
		for j := 0; j < nPoints; j++ {
			pheromone[i][j] = 1.0
		}
	}
	if rng == nil {
		rng = rand.New(rand.NewPCG(42, 0))
	}
	return &AntColony{
		Points:              points,
		NAnts:               nAnts,
		NIterations:         nIterations,
		NPoints:             nPoints,
		Pheromone:           pheromone,
		Alpha:               alpha,
		Beta:                beta,
		EvaporationRate:     evaporationRate,
		Q:                   q,
		BestPathLength:      math.Inf(1),
		BestDowntimeDays:    math.Inf(1),
		BestPathLenDowntime: math.Inf(1),
		Rnd:                 rng,
	}
}

func (ac *AntColony) UpdatePheromone(paths [][]int, objectives [][2]float64) {
	// Evaporation on all edges
	for i := 0; i < ac.NPoints; i++ {
		for j := 0; j < ac.NPoints; j++ {
			ac.Pheromone[i][j] = (1.0 - ac.EvaporationRate) * ac.Pheromone[i][j]
		}
	}

	// Deposit pheromone strictly on edges traversed by each ant
	for k := 0; k < len(paths); k++ {
		totalCost := objectives[k][0] + objectives[k][1]
		delta := ac.Q
		if totalCost > 1e-9 {
			delta = ac.Q / totalCost
		}

		path := paths[k]
		for t := 0; t < len(path)-1; t++ {
			u := path[t]
			v := path[t+1]
			ac.Pheromone[u][v] += delta
			ac.Pheromone[v][u] += delta
		}
	}

	// Avoid pheromone vanishing to exact 0
	for i := 0; i < ac.NPoints; i++ {
		for j := 0; j < ac.NPoints; j++ {
			if ac.Pheromone[i][j] < 1e-6 {
				ac.Pheromone[i][j] = 1e-6
			}
		}
	}
}

func (ac *AntColony) Optimize() {
	if ac.NPoints == 0 {
		ac.BestPath = []int{}
		ac.TurbineOrder = []string{}
		ac.BestPathLength = 0.0
		ac.BestDowntimeDays = 0.0
		ac.BestPathLenDowntime = 0.0
		return
	}

	if ac.NPoints == 1 {
		ac.BestPath = []int{0, 0}
		name := ac.Points[0].TurbineName
		if name == "" {
			name = "Point_0"
		}
		ac.TurbineOrder = []string{name, name}
		dist, dt := ObjectiveFunction(ac.BestPath, ac.Points)
		ac.BestPathLength = dist
		ac.BestDowntimeDays = dt
		ac.BestPathLenDowntime = dist + dt
		return
	}

	for iter := 0; iter < ac.NIterations; iter++ {
		paths := make([][]int, 0, ac.NAnts)
		objectives := make([][2]float64, 0, ac.NAnts)

		for ant := 0; ant < ac.NAnts; ant++ {
			visited := make([]bool, ac.NPoints)
			currentPoint := int(ac.Rnd.UintN(uint(ac.NPoints)))
			visited[currentPoint] = true
			path := []int{currentPoint}

			visitedCount := 1
			for visitedCount < ac.NPoints {
				unvisited := make([]int, 0, ac.NPoints-visitedCount)
				for i := 0; i < ac.NPoints; i++ {
					if !visited[i] {
						unvisited = append(unvisited, i)
					}
				}

				probs := make([]float64, len(unvisited))
				var sumProb float64

				currLatNorm := ac.Points[currentPoint].LatitudeNorm
				currLonNorm := ac.Points[currentPoint].LongitudeNorm
				currDtNorm := ac.Points[currentPoint].FaultDowntimeDaysNorm

				for i, unvis := range unvisited {
					pAlpha := math.Pow(ac.Pheromone[currentPoint][unvis], ac.Alpha)

					uLatNorm := ac.Points[unvis].LatitudeNorm
					uLonNorm := ac.Points[unvis].LongitudeNorm
					uDtNorm := ac.Points[unvis].FaultDowntimeDaysNorm

					dist := Distance([2]float64{currLatNorm, currLonNorm}, [2]float64{uLatNorm, uLonNorm})
					dtCost := DowntimeCost(currDtNorm, uDtNorm)
					combinedCost := dist + dtCost

					heuristic := math.Pow(1.0/(combinedCost+1e-10), ac.Beta)
					prob := pAlpha * heuristic
					probs[i] = prob
					sumProb += prob
				}

				var nextPoint int
				if sumProb > 0 && !math.IsNaN(sumProb) {
					r := ac.Rnd.Float64() * sumProb
					var cumSum float64
					nextPoint = unvisited[len(unvisited)-1]
					for i, prob := range probs {
						cumSum += prob
						if r <= cumSum {
							nextPoint = unvisited[i]
							break
						}
					}
				} else {
					nextPoint = unvisited[int(ac.Rnd.UintN(uint(len(unvisited))))]
				}

				path = append(path, nextPoint)
				visited[nextPoint] = true
				currentPoint = nextPoint
				visitedCount++
			}

			// Return to start
			path = append(path, path[0])

			dist, dtCost := ObjectiveFunction(path, ac.Points)
			paths = append(paths, path)
			objectives = append(objectives, [2]float64{dist, dtCost})

			totalCost := dist + dtCost
			if totalCost < ac.BestPathLenDowntime {
				ac.BestPath = path
				ac.BestPathLength = dist
				ac.BestDowntimeDays = dtCost
				ac.BestPathLenDowntime = totalCost

				order := make([]string, len(path))
				for i, idx := range path {
					order[i] = ac.Points[idx].TurbineName
				}
				ac.TurbineOrder = order
			}
		}

		ac.UpdatePheromone(paths, objectives)
	}
}

File: rust/src/algorithms/aco.rs
use rand::Rng;
use crate::models::TurbineFaultPoint;
use super::common::{distance, downtime_cost, objective_function};

#[derive(Debug, Clone)]
pub struct AntColony {
    pub points: Vec<TurbineFaultPoint>,
    pub n_ants: usize,
    pub n_iterations: usize,
    pub n_points: usize,
    pub pheromone: Vec<Vec<f64>>,
    pub alpha: f64,
    pub beta: f64,
    pub evaporation_rate: f64,
    pub q: f64,
    pub best_path: Vec<usize>,
    pub turbine_order: Vec<String>,
    pub best_path_length: f64,
    pub best_downtime_days: f64,
    pub best_path_len_downtime: f64,
}

impl AntColony {
    pub fn new(
        points: Vec<TurbineFaultPoint>,
        n_ants: usize,
        n_iterations: usize,
        alpha: f64,
        beta: f64,
        evaporation_rate: f64,
        q: f64,
    ) -> Self {
        let n_points = points.len();
        let pheromone = vec![vec![1.0; n_points]; n_points];

        Self {
            points,
            n_ants,
            n_iterations,
            n_points,
            pheromone,
            alpha,
            beta,
            evaporation_rate,
            q,
            best_path: Vec::new(),
            turbine_order: Vec::new(),
            best_path_length: f64::INFINITY,
            best_downtime_days: f64::INFINITY,
            best_path_len_downtime: f64::INFINITY,
        }
    }

    pub fn update_pheromone(&mut self, paths: &[Vec<usize>], objectives: &[(f64, f64)]) {
        // Evaporation on all edges
        for i in 0..self.n_points {
            for j in 0..self.n_points {
                self.pheromone[i][j] *= 1.0 - self.evaporation_rate;
            }
        }

        // Deposit pheromone strictly on edges traversed by each ant
        for (k, path) in paths.iter().enumerate() {
            let total_cost = objectives[k].0 + objectives[k].1;
            let delta = if total_cost > 1e-9 {
                self.q / total_cost
            } else {
                self.q
            };

            for t in 0..path.len() - 1 {
                let u = path[t];
                let v = path[t + 1];
                self.pheromone[u][v] += delta;
                self.pheromone[v][u] += delta;
            }
        }

        // Avoid pheromone vanishing to exact 0
        for i in 0..self.n_points {
            for j in 0..self.n_points {
                if self.pheromone[i][j] < 1e-6 {
                    self.pheromone[i][j] = 1e-6;
                }
            }
        }
    }

    pub fn optimize<R: Rng>(&mut self, rng: &mut R) {
        if self.n_points == 0 {
            self.best_path = Vec::new();
            self.turbine_order = Vec::new();
            self.best_path_length = 0.0;
            self.best_downtime_days = 0.0;
            self.best_path_len_downtime = 0.0;
            return;
        }

        if self.n_points == 1 {
            self.best_path = vec![0, 0];
            let name = if self.points[0].turbine_name.is_empty() {
                "Point_0".to_string()
            } else {
                self.points[0].turbine_name.clone()
            };
            self.turbine_order = vec![name.clone(), name];
            let (dist, dt) = objective_function(&self.best_path, &self.points);
            self.best_path_length = dist;
            self.best_downtime_days = dt;
            self.best_path_len_downtime = dist + dt;
            return;
        }

        let mut unvisited_buf = Vec::with_capacity(self.n_points);
        let mut probs_buf = Vec::with_capacity(self.n_points);

        for _iter in 0..self.n_iterations {
            let mut paths = Vec::with_capacity(self.n_ants);
            let mut objectives = Vec::with_capacity(self.n_ants);

            for _ant in 0..self.n_ants {
                let mut visited = vec![false; self.n_points];
                let current_point = rng.gen_range(0..self.n_points);
                visited[current_point] = true;
                let mut path = Vec::with_capacity(self.n_points + 1);
                path.push(current_point);

                let mut curr_node = current_point;
                let mut visited_count = 1;

                while visited_count < self.n_points {
                    unvisited_buf.clear();
                    for i in 0..self.n_points {
                        if !visited[i] {
                            unvisited_buf.push(i);
                        }
                    }

                    probs_buf.clear();
                    let mut sum_prob = 0.0;

                    let curr_lat = self.points[curr_node].latitude_norm;
                    let curr_lon = self.points[curr_node].longitude_norm;
                    let curr_dt = self.points[curr_node].fault_downtime_days_norm;

                    for &unvis in &unvisited_buf {
                        let p_alpha = self.pheromone[curr_node][unvis].powf(self.alpha);

                        let u_lat = self.points[unvis].latitude_norm;
                        let u_lon = self.points[unvis].longitude_norm;
                        let u_dt = self.points[unvis].fault_downtime_days_norm;

                        let dist = distance([curr_lat, curr_lon], [u_lat, u_lon]);
                        let dt_cost = downtime_cost(curr_dt, u_dt);
                        let combined_cost = dist + dt_cost;

                        let heuristic = (1.0 / (combined_cost + 1e-10)).powf(self.beta);
                        let prob = p_alpha * heuristic;
                        probs_buf.push(prob);
                        sum_prob += prob;
                    }

                    let next_point = if sum_prob > 0.0 && !sum_prob.is_nan() {
                        let r = rng.gen::<f64>() * sum_prob;
                        let mut cum_sum = 0.0;
                        let mut selected = unvisited_buf[unvisited_buf.len() - 1];
                        for (i, &prob) in probs_buf.iter().enumerate() {
                            cum_sum += prob;
                            if r <= cum_sum {
                                selected = unvisited_buf[i];
                                break;
                            }
                        }
                        selected
                    } else {
                        let idx = rng.gen_range(0..unvisited_buf.len());
                        unvisited_buf[idx]
                    };

                    path.push(next_point);
                    visited[next_point] = true;
                    curr_node = next_point;
                    visited_count += 1;
                }

                // Return to start
                path.push(path[0]);

                let (dist, dt_cost) = objective_function(&path, &self.points);
                let total_cost = dist + dt_cost;
                paths.push(path.clone());
                objectives.push((dist, dt_cost));

                if total_cost < self.best_path_len_downtime {
                    self.best_path = path.clone();
                    self.best_path_length = dist;
                    self.best_downtime_days = dt_cost;
                    self.best_path_len_downtime = total_cost;
                    self.turbine_order = path
                        .iter()
                        .map(|&idx| self.points[idx].turbine_name.clone())
                        .collect();
                }
            }

            self.update_pheromone(&paths, &objectives);
        }
    }
}
